kubetest/gke.go

```diff
@@ -0,0 +1,249 @@
+	gkeShape           = flag.String("gke-shape", `{"default":{"Nodes":3,"MachineType":"n1-standard-2"}}`, `(gke only) A JSON description of node pools to create. The node pool 'default' is required and used for initial cluster creation. All node pools are symmetric across zones, so the cluster total node count is {total nodes in --gke-shape} * {1 + (length of --gke-additional-zones)}. Example: '{"default":{"Nodes":999,"MachineType:":"n1-standard-1"},"heapster":{"Nodes":1, "MachineType":"n1-standard-8"}}`)
+	gkeNetwork         = flag.String("gke-network", "e2e", "(gke only) Cluster network. Must be set for GKE.")
+
+	zoneRe = regexp.MustCompile(`^.*/zones/([a-z0-9-]*)/instanceGroupManagers/([^/]+)$`)
+)
+
+type gkeNodePool struct {
+	Nodes       int
+	MachineType string
+}
+
+type gkeDeployer struct {
+	project         string
+	zone            string
+	additionalZones string
+	cluster         string
+	shape           map[string]gkeNodePool
+	network         string
+	version         string
+	kubecfg         string
+}
+
+var _ deployer = gkeDeployer{}
+
+func newGKE(project, zone string) (*gkeDeployer, error) {
+	g := &gkeDeployer{}
+
+	if project == "" {
+		return nil, fmt.Errorf("--gcp-project must be set")
+	}
+	g.project = project
+
+	if zone == "" {
+		return nil, fmt.Errorf("--gcp-zone must be set")
+	}
+	g.zone = zone
+
+	// TODO(zmerlynn): Additional enforcements on cluster name before we try to create it?
+	if *gkeCluster == "" {
+		return nil, fmt.Errorf("--gke-cluster must be set")
+	}
+	g.cluster = *gkeCluster
+
+	if *gkeNetwork == "" {
+		return nil, fmt.Errorf("--gke-network must be set")
+	}
+	g.network = *gkeNetwork
+
+	g.additionalZones = *gkeAdditionalZones
+
+	err := json.Unmarshal([]byte(*gkeShape), &g.shape)
+	if err != nil {
+		return nil, fmt.Errorf("--gke-shape must be valid JSON, unmarshal error: %v, JSON: %q", err, *gkeShape)
+	}
+	if _, ok := g.shape[defaultPool]; !ok {
+		return nil, fmt.Errorf("--gke-shape must include a node pool named 'default', found %q", *gkeShape)
+	}
+
+	// TODO(zmerlynn): The version should be plumbed through Extract
+	// or a separate flag rather than magic env variables.
+	g.version = os.Getenv("CLUSTER_API_VERSION")
+
+	// Override kubecfg to a temporary file rather than trashing the user's.
+	f, err := ioutil.TempFile("", "gke-kubecfg")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	kubecfg := f.Name()
+	if err := f.Chmod(0600); err != nil {
+		return nil, err
+	}
+	if err := os.Setenv("KUBECONFIG", kubecfg); err != nil {
+		return nil, err
+	}
+	g.kubecfg = kubecfg
+
+	// We want no KUBERNETES_PROVIDER, but to set
+	// KUBERNETES_CONFORMANCE_PROVIDER and
+	// KUBERNETES_CONFORMANCE_TEST. This prevents ginkgo-e2e.sh from
+	// using the cluster/gke functions.
+	//
+	// TODO(zmerlynn): This is gross.
+	if err := os.Unsetenv("KUBERNETES_PROVIDER"); err != nil {
+		return nil, err
+	}
+	if err := os.Setenv("KUBERNETES_CONFORMANCE_TEST", "yes"); err != nil {
```

Comment:
Use `pushEnv`, maybe write a similar `popEnv` for `KUBERNETES_PROVIDER`? Or actually we are switching `KUBERNETES_PROVIDER` to `--provider` and this will be fine. Or can they be flags?